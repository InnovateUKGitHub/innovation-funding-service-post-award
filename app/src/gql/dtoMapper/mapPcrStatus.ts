import { ProjectChangeRequestStatusChangeDto } from "@framework/dtos/pcrDtos";
import { mapToPCRStatus } from "@framework/mappers/pcr";
import { Clock } from "@framework/util/clock";

const clock = new Clock();

// on Acc_StatusChange__c

type PcrStatusNode = GQL.PartialNode<{
  Id: string;
  Acc_ProjectChangeRequest__c: GQL.Value<string>;
  Acc_PreviousProjectChangeRequestStatus__c: GQL.Value<string>;
  Acc_NewProjectChangeRequestStatus__c: GQL.Value<string>;
  CreatedDate: GQL.Value<string>;
  Acc_ExternalComment__c: GQL.Value<string>;
  Acc_CreatedByAlias__c: GQL.Value<string>;
  Acc_ParticipantVisibility__c: GQL.Value<boolean>;
}>;

type PcrStatusDtoMapping = ProjectChangeRequestStatusChangeDto & { pcrId: PcrId };

const mapper: GQL.DtoMapper<PcrStatusDtoMapping, PcrStatusNode, { roles?: SfRoles }> = {
  id(node) {
    return node?.Id ?? "";
  },
  projectChangeRequest(node) {
    return (node?.Acc_ProjectChangeRequest__c?.value ?? "unknown") as PcrId;
  },
  pcrId(node) {
    return (node?.Acc_ProjectChangeRequest__c?.value ?? "unknown") as PcrId;
  },
  createdBy(node) {
    return node?.Acc_CreatedByAlias__c?.value ?? "unknown";
  },
  createdDate(node) {
    return !!node?.CreatedDate?.value ? clock.parseRequiredSalesforceDateTime(node?.CreatedDate?.value) : new Date();
  },
  previousStatus(node) {
    return mapToPCRStatus(node?.Acc_PreviousProjectChangeRequestStatus__c?.value ?? "unknown");
  },
  previousStatusLabel(node) {
    return node?.Acc_PreviousProjectChangeRequestStatus__c?.value ?? "unknown";
  },
  newStatus(node) {
    return mapToPCRStatus(node?.Acc_NewProjectChangeRequestStatus__c?.value ?? "unknown");
  },
  newStatusLabel(node) {
    return node?.Acc_NewProjectChangeRequestStatus__c?.value ?? "unknown";
  },

  comments(node, additionalData) {
    /**
     * Required fields include:
     * Acc_ParticipantVisibility__c
     * Acc_ExternalComment__c
     * roles from Acc_Project__c
     */
    if (additionalData?.roles?.isMo || (additionalData?.roles?.isPm && node?.Acc_ParticipantVisibility__c?.value)) {
      return node?.Acc_ExternalComment__c?.value ?? null;
    }
    return null;
  },
  participantVisibility(node) {
    return node?.Acc_ParticipantVisibility__c?.value ?? false;
  },
};

type PcrStatusAdditionalData<TPickList extends string> = AdditionalDataType<
  TPickList,
  [
    ["comments", "roles", SfRoles], // get from Acc_Project__c
  ]
>;

/**
 * Maps a specified PcrStatus Node from a GQL query to a slice of
 * the PcrStatusDto to ensure consistency and compatibility in the application
 */
export function mapToPcrStatusDto<T extends PcrStatusNode, PickList extends keyof PcrStatusDtoMapping>(
  node: T,
  pickList: PickList[],
  additionalData: PcrStatusAdditionalData<PickList>,
): Pick<PcrStatusDtoMapping, PickList> {
  return pickList.reduce(
    (dto, field) => {
      dto[field] = mapper[field](node, additionalData);
      return dto;
    },
    {} as Pick<PcrStatusDtoMapping, PickList>,
  );
}

/**
 * Maps PcrStatus Edge to array of PcrStatus dtos.
 */
export function mapToPcrStatusDtoArray<
  T extends ReadonlyArray<GQL.Maybe<{ node: PcrStatusNode }>> | null,
  PickList extends keyof PcrStatusDtoMapping,
>(
  edges: T,
  pickList: PickList[],
  additionalData: PcrStatusAdditionalData<PickList>,
): Pick<PcrStatusDtoMapping, PickList>[] {
  return (
    edges?.map(node => {
      return mapToPcrStatusDto(node?.node ?? null, pickList, additionalData);
    }) ?? []
  );
}
