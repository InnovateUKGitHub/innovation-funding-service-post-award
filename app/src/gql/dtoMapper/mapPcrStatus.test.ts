import { mapToPcrStatusDtoArray } from "./mapPcrStatus";

const edges = [
  {
    node: {
      Id: "a0F26000009akjeEAA",
      Acc_ProjectChangeRequest__c: {
        value: "a0G26000007KqPTEA0",
      },
      Acc_PreviousProjectChangeRequestStatus__c: {
        value: "Submitted to Monitoring Officer",
      },
      Acc_ExternalComment__c: {
        value: "all good",
      },
      CreatedDate: {
        value: "2023-05-24T09:46:09.000Z",
      },
      Acc_CreatedByAlias__c: {
        value: "Innovate UK",
      },
      Acc_NewProjectChangeRequestStatus__c: {
        value: "Submitted to Innovate UK",
      },
    },
  },
  {
    node: {
      Id: "a0F26000009akjZEAQ",
      Acc_ProjectChangeRequest__c: {
        value: "a0G26000007KqPTEA0",
      },
      Acc_PreviousProjectChangeRequestStatus__c: {
        value: "Draft",
      },
      Acc_ExternalComment__c: {
        value: null,
      },
      CreatedDate: {
        value: "2023-05-24T09:45:11.000Z",
      },
      Acc_CreatedByAlias__c: {
        value: "Innovate UK",
      },
      Acc_NewProjectChangeRequestStatus__c: {
        value: "Submitted to Monitoring Officer",
      },
    },
  },
  {
    node: {
      Id: "a0F26000009akjFEAQ",
      Acc_ProjectChangeRequest__c: {
        value: "a0G26000007KqPTEA0",
      },
      Acc_PreviousProjectChangeRequestStatus__c: {
        value: null,
      },
      Acc_ExternalComment__c: {
        value: null,
      },
      CreatedDate: {
        value: "2023-05-24T09:29:49.000Z",
      },
      Acc_CreatedByAlias__c: {
        value: "Innovate UK",
      },
      Acc_NewProjectChangeRequestStatus__c: {
        value: "Draft",
      },
    },
  },
];

describe("mapPcrStatusDtoArray", () => {
  it("should map the node to the projectDto structure", () => {
    expect(
      mapToPcrStatusDtoArray(
        edges,
        [
          "id",
          "pcrId",
          "createdBy",
          "createdDate",
          "newStatus",
          "previousStatus",
          "newStatusLabel",
          "comments",
          "previousStatusLabel",
        ],
        { roles: { isMo: true, isFc: false, isPm: true, isAssociate: false } },
      ),
    ).toMatchSnapshot();
  });
});
