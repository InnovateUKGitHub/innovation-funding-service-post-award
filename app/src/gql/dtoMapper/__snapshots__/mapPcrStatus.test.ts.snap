// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapPcrStatusDtoArray should map the node to the projectDto structure 1`] = `
Array [
  Object {
    "comments": "all good",
    "createdBy": "Innovate UK",
    "createdDate": 2023-05-24T09:46:09.000Z,
    "id": "a0F26000009akjeEAA",
    "newStatus": 12,
    "newStatusLabel": "Submitted to Innovate UK",
    "pcrId": "a0G26000007KqPTEA0",
    "previousStatus": 2,
    "previousStatusLabel": "Submitted to Monitoring Officer",
  },
  Object {
    "comments": null,
    "createdBy": "Innovate UK",
    "createdDate": 2023-05-24T09:45:11.000Z,
    "id": "a0F26000009akjZEAQ",
    "newStatus": 2,
    "newStatusLabel": "Submitted to Monitoring Officer",
    "pcrId": "a0G26000007KqPTEA0",
    "previousStatus": 1,
    "previousStatusLabel": "Draft",
  },
  Object {
    "comments": null,
    "createdBy": "Innovate UK",
    "createdDate": 2023-05-24T09:29:49.000Z,
    "id": "a0F26000009akjFEAQ",
    "newStatus": 1,
    "newStatusLabel": "Draft",
    "pcrId": "a0G26000007KqPTEA0",
    "previousStatus": 0,
    "previousStatusLabel": "unknown",
  },
]
`;
